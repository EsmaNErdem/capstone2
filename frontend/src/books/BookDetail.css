.BookDetail {
    margin: 20px;
  }

  .BookDetailContainer {
    border: 0;
  }

  .BookCoverPaper {
    height: 100%;
    display: flex;
    align-items: center;    
    justify-content: center;

  }
  
  .BookCoverImage {
    width: auto;
    max-width: 3750px;
    height: 100%;
    max-height: 465px;
    object-fit: fit;
  }
  
  .BookDetailPaper {
    padding: 20px;
  }
  
  .BookSummary {
    margin-bottom: 20px;
  }
  
  .CardFooter {
    display: flex;
    justify-content:space-around;
    align-items: center;
    margin-top: 10px;
  }
  
  .LikeSection,
  .ReviewSection {
    display: flex;
    align-items: center;
  }
  
  .ReviewButton {
    color: #000; 
  }
  
  .HasReviews {
    color: yellowgreen; 
  }

  .Review-input {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    background-color: #fff; /* Change to your desired background color */
    box-shadow: 0px 0px 24px rgba(0, 0, 0, 0.2); /* Change to your desired box shadow */
    padding: 16px;
    width: 90%;
    max-width: 550px;

    @media (min-width: 600px) {
        width: 70%;
    }

    @media (min-width: 960px) {
        width: 60%;
    }

    @media (min-width: 1280px) {
        width: 50%;
    }
  }
  