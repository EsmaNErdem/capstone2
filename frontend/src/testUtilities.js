import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
