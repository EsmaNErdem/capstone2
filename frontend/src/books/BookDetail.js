import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import useBookLike from "../hooks/useBookLike";
import useReviewAdd from "../hooks/useReviewAdd";
import useReviewDelete from "../hooks/useReviewDelete";
import BookClubApi from "../api";
import BookReviewAddForm from "./BookReviewAddForm"
import ReviewFilterForm from "../reviews/ReviewFilterForm";
import ReviewCard from "../reviews/ReviewCard"
import Loading from "../utilities/Loading";
import Alert from "../utilities/Alert";
import { Box, Grid, Paper, List, IconButton, Modal }from '@mui/material';
import FavoriteIcon from '@mui/icons-material/Favorite';
import AddCommentIcon from '@mui/icons-material/AddComment';
import './BookDetail.css';

/**
 * Displays detailed info about a book
 * The `BookDetail` component displays a book detail info fetched from the Google Books API,
 * along with database reviews and likes. It supports both infinite scrolling for loading
 * more book related reviews and filter book reviews based on user input.
 * 
 * - API call loads book data when the component mounts with getBook and when the filter changes to load updated review list with getAllReviewsByBook from BookClubApi
 * - Send user like on a book. If fails, shows error to user
 * - This component is designed for the "/books/:id" route.
 * 
 * - Routes ==> BookCard, BookSearchForm
 */
const BookDetail = () => {
    console.debug("BookDetail");

    const { id } = useParams();

    const [loading, setLoading] = useState(true)
    const [book, setBook] = useState(null);
    const [bookReviews, setBookReviews] = useState([]);
    const [reviewFormOpen, setReviewFormOpen] = useState(false);
    const [error, setError] = useState(null);

    const { error: deleteError, deleteBookReview } = useReviewDelete(setBookReviews);
    
    /**
     * Fetches the book data from the Google Books API and related reviews from database, then sets the bookReviews state.
    */  
   useEffect(function getBookOnMount() {
       console.debug("BookDetail useEffect getBookOnMount");
       
       // Make an API call to get book data and book related reviews
       const getBookData =  async () => {
            setError(null);
            try {
                const bookData = await BookClubApi.getBook(id);    
                setBook(bookData)
                getFilteredBookReviews({ sortBy: "date" })
            } catch (e) {
                console.error("BookDetail useEffect API call data loading error:", e);
                setError("An error occurred while fetching book data.");
            }
            
            setLoading(false)
        }
        
        // set loading to true while async getCurrentUser runs; once the
        // data is fetched (or even if an error happens!), this will be set back
        // to true to control the spinner.
        setLoading(true)
        getBookData()
    }, [id]);
    
    // With form submission, make an API call and update bookReviews state
    const getFilteredBookReviews = async (data) => {
        try {
            setBookReviews([])
            const reviews = await BookClubApi.getAllReviewsByBook(id, data)
            setBookReviews(reviews)
        } catch (e) {
            console.error("BookDetail get filtered book reviews", e);
            setError("An error occurred while fetching filtered reviews.");
        }
    }

    // Opens Modal to display review text input 
    const openReviewAddForm = async () => {
        setReviewFormOpen(true)
    }
    
    // Closes  Modal to display review text input 
    const closeReviewAddForm = async () => {
        setReviewFormOpen(false)
    }

    const databaseBookData = {
        id,
        title: book?.title,
        author: book?.author,
        publisher: book?.publisher || undefined,
        description: book?.description || undefined,
        category: book?.category || undefined,
        cover: book?.cover || undefined,
    }

    // make sure to wait untill book detail data loads
    const { liked, likes, error: likeError, handleLikeBook } = useBookLike(
        id,
        +book?.bookLikeCount,
        databaseBookData
    );
        
    const { error: addError, addBookReview } = useReviewAdd(
        setBookReviews, 
        databaseBookData, 
        setReviewFormOpen
    );

    if(loading || !book) return <Loading />;
    
    return (
        <div className="BookDetail">
            <Box sx={{ border: 0 }}>
                <Grid container spacing={3} className="BookDetailContainer" style={{ display: 'flex', justifyContent: 'center' }}>
                    <Grid item xs={12} md={4} lg={6}>
                        <Paper className="BookCoverPaper">
                            <img
                            alt={book.title}
                            src={book.cover}
                            className="BookCoverImage"
                            />
                        </Paper>
                    </Grid>
                    <Grid item xs={12} md={6}>
                        <Paper className="BookDetailPaper">
                            <h1 data-testid="book-title">{book.title}</h1>
                            <p>by {book.author}</p>
                            {book?.publisher && <p>Publisher {book.publisher}</p>}
                            {book?.categories && (
                            <div>
                                <strong>Categories: </strong>
                                {book.categories.map((category, index) => (
                                <span key={index}>
                                    {category}
                                    {index !== book?.categories.length - 1 ? ", " : ""}
                                </span>
                                ))}
                            </div>
                            )}
                            <p className="BookSummary">{removeHtmlTags(book?.description)}</p>
                            <div className="CardFooter">
                                <div className="LikeSection">
                                    <span>{likes ? likes : 0}</span>
                                    <IconButton
                                        onClick={handleLikeBook}
                                        color={liked ? "error" : "default"}
                                    >
                                    <FavoriteIcon data-testid="like-button" />
                                    </IconButton>
                                </div>
                                <div className="ReviewSection">
                                    <span>{bookReviews ? bookReviews.length : 0}</span>
                                    <IconButton
                                        data-testid="add-review-toggle"
                                        onClick={openReviewAddForm}
                                        className="ReviewButton"
                                        style={{ color: bookReviews.length > 0 ? "yellowgreen" : ""}}
                                        >
                                    <AddCommentIcon data-testid="review-button" />
                                    </IconButton>
                                </div>
                            </div>
                        </Paper>
                    </Grid>
                </Grid>
            </Box>
            
            {(likeError || deleteError || addError)&& <Alert type="danger" messages={[likeError || deleteError || addError]} />}
            {error && <Alert type="danger" messages={[error]} />}
            
            <Modal open={reviewFormOpen} onClose={closeReviewAddForm}>
                <Box className="Review-input" >
                    <BookReviewAddForm addReviews={addBookReview} rowCount={6} close={true} closeModal={closeReviewAddForm}/>
                </Box>
            </Modal>

            <ReviewFilterForm applyFilters={getFilteredBookReviews} prompts={["username"]} initialValue={{ sortBy: "date" }} />
            <List>
                {!bookReviews && <Loading />}
                {bookReviews.map(({ reviewId, review, username, userImg, date, reviewLikeCount }) => 
                        <ReviewCard 
                            key={reviewId}
                            reviewId={reviewId}
                            review={review}
                            username={username}
                            userImg={userImg}
                            date={date}
                            reviewLikeCount={+reviewLikeCount}
                            deleteReview={deleteBookReview}
                        />
                    )}
            </List>
        </div>
    )
}

export default BookDetail;

const removeHtmlTags = (htmlString) => {
    var doc = new DOMParser().parseFromString(htmlString, 'text/html');
    return doc.body.textContent || "";
}