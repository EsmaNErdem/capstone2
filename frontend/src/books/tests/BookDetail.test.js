import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import BookDetail from '../BookDetail';
import BookClubApi from '../../api';

const mockBook = {
    id: '1',
    title: 'Book1',
    author: 'Author1',
    publisher: 'Publisher1',
    description: 'Description1',
    category: 'Category1',
    cover: 'Cover1',
    bookLikeCount: 5,
}

const mockReviews = [{
        reviewId: 1,
        review: "Review1",
        username: "Test1", 
        userImg: "UserImg1",
        date: "12/18/2023"
    }, 
    {
        reviewId: 2,
        review: "Review2",
        username: "Test2", 
        userImg: "UserImg2",
        date: "12/18/2023"
    }]

jest.mock("../../api", () => ({
    __esModule: true,
    default: {
        getBook: jest.fn(),
        getAllReviewsByBook: jest.fn(),
    },
}));

afterEach(() => {
jest.clearAllMocks();
});

  
test('renders without crashing', async () => {
    BookClubApi.getBook.mockImplementationOnce(() => (mockBook));
    BookClubApi.getAllReviewsByBook.mockImplementationOnce(() => (mockReviews));
  
    await act(async () => {
      render(
        <MemoryRouter>
          <UserProvider>
            <BookDetail />
          </UserProvider>
        </MemoryRouter>
      );
    });
  
  });
  
  test('it renders and matches with snapshot', async () => {
    BookClubApi.getBook.mockImplementationOnce(() => (mockBook));
    BookClubApi.getAllReviewsByBook.mockImplementationOnce(() => (mockReviews));

    await act(async () => {
      ({ asFragment, getByTestId } = render(
        <MemoryRouter>
          <UserProvider>
            <BookDetail />
          </UserProvider>
        </MemoryRouter>
      ));
    });
  
    expect(asFragment()).toMatchSnapshot();

    const reviewAdd = getByTestId("add-review-toggle")
    fireEvent.click(reviewAdd);
    
    await waitFor(() => {
        expect(asFragment()).toMatchSnapshot();
    });
    
});

test('displays book details and reviews', async () => {
    BookClubApi.getBook.mockImplementationOnce(() => (mockBook));
    BookClubApi.getAllReviewsByBook.mockImplementationOnce(() => (mockReviews));
    
    await act(async () => {
        ({ getByText, getByTestId, getAllByTestId, getByLabelText, queryByLabelText, getByPlaceholderText } = render(
            <MemoryRouter>
          <UserProvider>
            <BookDetail />
          </UserProvider>
        </MemoryRouter>
      ));
    });
    
    expect(getByText('Book1')).toBeInTheDocument();
    expect(getByText('Review1')).toBeInTheDocument();
    expect(getAllByTestId('like-button')[0]).toBeInTheDocument();
    expect(queryByLabelText('Add a Review')).not.toBeInTheDocument();
    
    const reviewAdd = getByTestId("add-review-toggle")
    fireEvent.click(reviewAdd);
    
    await waitFor(() => {
        expect(getByLabelText('Add a Review')).toBeInTheDocument();
    });
  
    expect(getByPlaceholderText('Review Owner')).toBeInTheDocument();
});
  
  