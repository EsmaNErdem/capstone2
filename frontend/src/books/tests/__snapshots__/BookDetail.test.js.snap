// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="BookDetail"
  >
    <div
      class="MuiBox-root css-1k9h2vf"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 BookDetailContainer css-zow5z4-MuiGrid-root"
        style="display: flex; justify-content: center;"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 MuiGrid-grid-lg-6 css-gg3qly-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 BookCoverPaper css-1ps6pg7-MuiPaper-root"
          >
            <img
              alt="Book1"
              class="BookCoverImage"
              src="Cover1"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 BookDetailPaper css-1ps6pg7-MuiPaper-root"
          >
            <h1
              data-testid="book-title"
            >
              Book1
            </h1>
            <p>
              by Author1
            </p>
            <p>
              Publisher Publisher1
            </p>
            <p
              class="BookSummary"
            >
              Description1
            </p>
            <div
              class="CardFooter"
            >
              <div
                class="LikeSection"
              >
                <span>
                  5
                </span>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="like-button"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <div
                class="ReviewSection"
              >
                <span>
                  2
                </span>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium ReviewButton css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  data-testid="add-review-toggle"
                  style="color: yellowgreen;"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="review-button"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM17 11h-4v4h-2v-4H7V9h4V5h2v4h4v2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="FilterBox"
    >
      <form
        class="FilterForm"
      >
        <input
          placeholder="Review Owner"
          value=""
        />
        <select
          name="sortBy"
        >
          <option
            value=""
          >
            Sort by
          </option>
          <option
            value="date"
          >
            Sort by Date ↑
          </option>
          <option
            value="user"
          >
            User by Alphabet ↓
          </option>
          <option
            value="popular"
          >
            Sort by Popularity ↑
          </option>
        </select>
        <button
          class="SubmitButton"
          title="Apply Filters"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-filter "
            data-icon="filter"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.9 54.9C10.5 40.9 24.5 32 40 32H472c15.5 0 29.5 8.9 36.1 22.9s4.6 30.5-5.2 42.5L320 320.9V448c0 12.1-6.8 23.2-17.7 28.6s-23.8 4.3-33.5-3l-64-48c-8.1-6-12.8-15.5-12.8-25.6V320.9L9 97.3C-.7 85.4-2.8 68.8 3.9 54.9z"
              fill="currentColor"
            />
          </svg>
        </button>
      </form>
    </div>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      >
        <a
          data-testid="review-card-profile-link"
          href="/profile/Test1"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1fy7wd8-MuiAvatar-root"
          >
            <img
              alt="Test1"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="UserImg1"
            />
          </div>
        </a>
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Review1
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
          >
            Test1 • 12/18/2023 12:00:00 AM
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="review-card-review-like"
          tabindex="0"
          type="button"
        >
          <span>
            NaN
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </li>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <li
        class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      >
        <a
          data-testid="review-card-profile-link"
          href="/profile/Test2"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1fy7wd8-MuiAvatar-root"
          >
            <img
              alt="Test2"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="UserImg2"
            />
          </div>
        </a>
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Review2
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
          >
            Test2 • 12/18/2023 12:00:00 AM
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="review-card-review-like"
          tabindex="0"
          type="button"
        >
          <span>
            NaN
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </li>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
    </ul>
  </div>
</DocumentFragment>
`;

exports[`it renders and matches with snapshot 2`] = `
<DocumentFragment>
  <div
    class="BookDetail"
  >
    <div
      class="MuiBox-root css-1k9h2vf"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 BookDetailContainer css-zow5z4-MuiGrid-root"
        style="display: flex; justify-content: center;"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 MuiGrid-grid-lg-6 css-gg3qly-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 BookCoverPaper css-1ps6pg7-MuiPaper-root"
          >
            <img
              alt="Book1"
              class="BookCoverImage"
              src="Cover1"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 BookDetailPaper css-1ps6pg7-MuiPaper-root"
          >
            <h1
              data-testid="book-title"
            >
              Book1
            </h1>
            <p>
              by Author1
            </p>
            <p>
              Publisher Publisher1
            </p>
            <p
              class="BookSummary"
            >
              Description1
            </p>
            <div
              class="CardFooter"
            >
              <div
                class="LikeSection"
              >
                <span>
                  5
                </span>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="like-button"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <div
                class="ReviewSection"
              >
                <span>
                  2
                </span>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium ReviewButton css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  data-testid="add-review-toggle"
                  style="color: yellowgreen;"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="review-button"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM17 11h-4v4h-2v-4H7V9h4V5h2v4h4v2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="FilterBox"
    >
      <form
        class="FilterForm"
      >
        <input
          placeholder="Review Owner"
          value=""
        />
        <select
          name="sortBy"
        >
          <option
            value=""
          >
            Sort by
          </option>
          <option
            value="date"
          >
            Sort by Date ↑
          </option>
          <option
            value="user"
          >
            User by Alphabet ↓
          </option>
          <option
            value="popular"
          >
            Sort by Popularity ↑
          </option>
        </select>
        <button
          class="SubmitButton"
          title="Apply Filters"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-filter "
            data-icon="filter"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.9 54.9C10.5 40.9 24.5 32 40 32H472c15.5 0 29.5 8.9 36.1 22.9s4.6 30.5-5.2 42.5L320 320.9V448c0 12.1-6.8 23.2-17.7 28.6s-23.8 4.3-33.5-3l-64-48c-8.1-6-12.8-15.5-12.8-25.6V320.9L9 97.3C-.7 85.4-2.8 68.8 3.9 54.9z"
              fill="currentColor"
            />
          </svg>
        </button>
      </form>
    </div>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      >
        <a
          data-testid="review-card-profile-link"
          href="/profile/Test1"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1fy7wd8-MuiAvatar-root"
          >
            <img
              alt="Test1"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="UserImg1"
            />
          </div>
        </a>
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Review1
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
          >
            Test1 • 12/18/2023 12:00:00 AM
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="review-card-review-like"
          tabindex="0"
          type="button"
        >
          <span>
            NaN
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </li>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <li
        class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      >
        <a
          data-testid="review-card-profile-link"
          href="/profile/Test2"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1fy7wd8-MuiAvatar-root"
          >
            <img
              alt="Test2"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="UserImg2"
            />
          </div>
        </a>
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Review2
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
          >
            Test2 • 12/18/2023 12:00:00 AM
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="review-card-review-like"
          tabindex="0"
          type="button"
        >
          <span>
            NaN
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </li>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
    </ul>
  </div>
</DocumentFragment>
`;
